source/Mlos.Python/mlos/Optimizers/OptimizerBase.py

```diff
@@ -91,10 +94,22 @@ def optimum(self, optimum_definition: OptimumDefinition = OptimumDefinition.BEST
         if not len(features_df.index):
             raise ValueError("Can't compute optimum before registering any observations.")
 
+        if context is not None and optimum_definition != OptimumDefinition.BEST_PREDICTED_WITHIN_CONTEXT:
+                raise ValueError(f"{optimum_definition} not supported if context is provided.")
+
         if optimum_definition == OptimumDefinition.BEST_OBSERVATION:
             return self._best_observation_optimum(features_df=features_df, objectives_df=objectives_df)
+        elif optimum_definition == OptimumDefinition.BEST_PREDICTED_WITHIN_CONTEXT:

```

Comment:
Can we make context orthogonal here? If the user supplies context, they get an optimum for that context, otherwise they get an optimum across all contexts?

And for the second case, should we return a config that's robust across contexts or one that was the best ever? I guess here you chose a speculative config for a given context. Should we make it explicit in the name?

Reply:
in what setting would a user want an optimum across contexts? And you're saying you'd take the maximum over contexts? Why the max and not the minimax (i.e. the one that's the best in the worst context) or the mean?

Reply:
One example I could think of is to use the medium/mean best across all workload types in order to encode a sensible "robust" default when the instance optimized is not known/available/working well.

Reply:
yes indeed, that might make sense. The maximum seemed just an odd choice. Doing mean/median it might make sense to use just historical data.
If we do predictions, would you want to average over only the context we've seen? So we run an optimization for each context and select the one with best mean performance?
You might also want to integrate over the whole space but then the distribution over contexts is a bit tricky, so using the empirical distribution there might make sense.

Reply:
So this is still a speculative context. In other words, the optimizer might suggest a config that had never been tried. Can we make it explicit in the name or by supplying another parameter? 

I think it might be a useful parameter for all other definitions of optimum...